packages/stan/README.md

```diff
@@ -0,0 +1,208 @@
+
+## Getting started
+
+Stan is based on the concept of "atoms". Atoms are discrete state units and your application state is a tree of atoms that depend on each other.
+
+### Creating basic atoms
+
+Let's create some basic atoms:
+
+```js
+import { createAtom } from '@wordpress/stan';
+
+// Creates an atom with an initial value of 1.
+// The value can be of any type.
+const counter = createAtom( 1 );
+```
+
+### Manipulating atoms
+
+In this example we created an atom that can hold a counter that starts with `1`. 
+But in order to manipulate that data, we need to create an "instance" of that atom.
+We do so by adding the atom to a registry of atoms. 
+The registry will then become a container of all instanciated atoms.
+
+```js
+import { createAtomRegistry } from '@wordpress/stan';
+
+const registry = createAtomRegistry();
+
+// this creates the "counter" atom or retrieves it if it's already in the registry.
+const counterInstance = registry.getAtom( counter );
+
+// Manipulate the atom.
+console.log( counterInstance.get() ); //  prints 1.
+
+// Modify the value of the counter
+counterInstance.set( 10 );
+console.log( counterInstance.get() ); //  prints 10.
+```
+
+### Subscribing to changes
+
+Each atom is an observable to which we can subscribe:
+
+```js
+counterInstance.subscribe( () => {
+    console.log( conterInstance.get() );
+} );
+
+counterInstance.set( 2 ); // prints 2.
+counterInstance.set( 4 ); // prints 4.
+```
+
+### Derived atoms
+
+Atoms can also derive their value based on other atoms. We call these "derived atoms".
+
+```js
+import { createAtom, createDerivedAtom } from '@wordpress/stan';
+
+const counter1 = createAtom( 1 );
+const counter2 = createAtom( 2 );
+const sum = createDerivedAtom(
+    ( get ) => get( counter1 ) + get( counter2 )
+);
+```
+
+In the example above, we create two simple counter atoms and third derived "sum" atom which value is the sum of both counters. Let's see how we can interact this atom. 
+
+So just like any other kind of atoms, we need an instance to manipulate it. Note also that adding an atom to a registry, automatically adds all its dependencies to the registry and creates instances for them if not already there.
+
+```js
+// Retrieve the sum instance and adds the counter1 and counter2 atoms to the registry as well
+const sumInstance = registry.getAtom( sum );
+```
+
+One important thing to note here as well is that atoms are "lazy", which means unless someone subscribes to their changes, they won't bother computing their state. This is an important property of atoms for performance reasons. 
+
+```js
+// No one has subscribed to the sum instance yet, its value is "null"
+console.log( sumInstance.get() ); // prints null.
```

Comment:
This looks wrong. If I have a React component that wants to render the `sumInstance.get()` value, then even the initial render should print `3`, not `null`.

The subscription will be created only after the initial render, in a `useEffect` or `componentDidMount` callback. It can't be created anywhere else, because it's a side effect.

This is true no matter how exactly the React bindings are implemented (`@wordpress/data` adapter or anything else).

Reply:
> This looks wrong. If I have a React component that wants to render the sumInstance.get() value, then even the initial render should print 3, not null

This is already handled properly in useSelect, we can compute the initial value before the subscription if we want. I may add a more direct way to do that with atoms: make the "resolve" a public API of the atom instance but it's very important that the atoms stay lazy. If I don't use a selector (which becomes an atom), I don't want it to keep updating its value and consuming CPU.

Reply:
This is a good remark though, and I need to think more about how to solve this in a better way: having the laziness while allowing sync resolution for react hooks.

Reply:
> This is already handled properly in useSelect, we can compute the initial value before the subscription if we want.

I tested this with the following example:
```js
const count = createAtom( 1 );
const double = createDerivedAtom( ( get ) => get( count ) * 2 );

const getDouble = ( get ) => () => get( double );

registerAtomicStore( 'standa', {
	atoms: { double },
	selectors: { getDouble },
} );

function UI() {
	const dbl = useSelect( ( select ) => select( 'standa' ).getDouble() );
	return React.createElement( 'div', null, dbl );
}

const rendered = ReactDOMServer.renderToString(
	React.createElement( UI, null )
);

console.log( rendered );
```
In the end, the `double` atom is really resolved even at the time of the initial render and the example correctly renders `<div>2</div>`.

But the reason _why_ it is resolved is not great: the atoms in the atomic store are instantiated in the `data` package's own atom registry, and that registry has an `onAdd` callback that immediately subscribes to the new atom's changes and relays them to the `globalListener`. I.e., every atom has at least one subscription during its entire lifetime. That ensures that the atom is resolved in the initial render, but at the same time completely defeats the lazy optimizations.

After removing the `onAdd` and `onDelete` callbacks from the `createAtomRegistry` call (with SSR, we don't need any subscriptions anyway), the resolution of `double` is not done. The selector returns `null` on initial render.

Reply:
> This is a good remark though, and I need to think more about how to solve this in a better way: having the laziness while allowing sync resolution for react hooks.

The selector value should be computed when calling `.get()`. Calling the getter is a strong-enough signal that someone is interested in the value and that it's worth computing.

Reply:
@jsnajdr it's not that simple, resolving might trigger effects (async atoms) in which case, we shouldn't resolve in render phase, I think it's more subtle and we need to figure out a way to declaratively separate async atoms from sync ones.

Reply:
and yes, I agree the global listener is not a great idea, it's a temporary measure that I put in place for backward compatibility reasons. I'll have to think more about that and probably only subscribe to the "root" atoms there.

Reply:
Ok, I've tried the approach you're suggestion in 0c6cd78 
it seems to work well but there's something wrong with the e2e tests. It doesn't feel related to my commit but I don't know.

Reply:
Tests aare back to normal and we're gaining 10ms when typing in this PR (this is huge)

Reply:
> resolving might trigger effects (async atoms) in which case, we shouldn't resolve in render phase

It's interesting that both Recoil and Jotai run the resolvers, both sync and async ones, during `get`. I.e., during the render phase, firing the side effects right away.

That suggests that running the resolver or any side effect during render is not bad per-se. We only need to avoid doing certain things that would make it bad:
- the resolver shouldn't do any sync state updates, like setting the `isResolving` flag and triggering a `setState` by that
- the resolver shouldn't allocate any resources (e.g., make subscriptions) because they won't be cleaned up in case of aborted render
- the resolver should be resilient against running twice

Looking at Apollo Client, I see they execute the query in two phases: `queryData.execute()` during render, and then `queryData.afterExecute()` in a `useEffect` hook. Only the latter can allocate something and be sure that it will be cleaned up.

To implement a good React binding, maybe exposing only the simple `get` primitive and nothing else is not enough? The current `get` does two tasks at once: "synchronously reading the value" and "triggering the resolution". There can easily be two lower-level primitives that do that separately.

Reply:
> To implement a good React binding, maybe exposing only the simple get primitive and nothing else is not enough? The current get does two tasks at once: "synchronously reading the value" and "triggering the resolution". There can easily be two lower-level primitives that do that separately.

Yes, I agree, I think there's value in "execute" and "afterExecute" to handle things like resolution state... but I believe we can start with the current API and see later. The important thing is to ensure `@wordpress/stan` remains an internal package (like icons and interface) where we're able to make breaking changes. (I'll update the building to do that)

Reply:
> but I believe we can start with the current API and see later.

My perspective is that I'd like to implement something like React Query on top of the `stan` store. That means maintaining mainly collections rather than individual atomic values, performing CRUD operations on them, acknowledging that these operations can fail and having first-class support for error handling, having control over the cache ("invalidating resolution" in the language of `@wordpress/data`)...

Recoil or Jotai provide some basic support for that, but I feel I'd quickly hit a wall when trying to implement some of the more complex requirements.

Reply:
yes, that's definitely where we'll see the most complex use-cases :) (basically, how to write core-data with stan).
I believe as is right now, it will provide a lot of good for performance for things like block-editor store but handling the complexities of something like `core-data` is still something that need to be tested.